### PR TITLE
Configure buffer times

### PR DESCRIPTION
**Pull request by Hlebanja**

### CODE REVIEW

File: Database/Buffer/Backend/index.ts [hunk 1]
```diff
@@ -10,29 +10,33 @@ import { Context } from "./Context"
...
 		)
 	}
-
-	async configuration(request: Request): Promise<void> {
+	async configure(request: Request): Promise<void> {
 		//TODO: Review
+		const snooze = +(request.headers.get("secondsBetweenArchives") ?? NaN)
+		this.snooze = this.snooze ?? (Number.isNaN(snooze) ? undefined : snooze)
+		const archiveTime = +(request.headers.get("secondsInBuffer") ?? NaN)
+		this.archiveTime = this.archiveTime ?? (Number.isNaN(archiveTime) ? undefined : archiveTime)
```
Comment (marcushanses) [inline Database/Buffer/Backend/index.ts:39]: These are not the actual headers. Headers are in snake case not in camel.

File: Database/Buffer/Backend/index.ts [hunk 2]
```diff
@@ -61,7 +64,6 @@ export class Backend {
 			return await archivist.reconcile(archiveThreshold)
 		})
 		// TODO: Should be set by some config
-		const snooze = now + archiveTime - 15 * 1000
-		await this.state.storage.setAlarm(snooze)
+		await this.state.storage.setAlarm(this.snooze)
```
Comment (marcushanses) [inline Database/Buffer/Backend/index.ts:67]: this.snooze is in seconds but the setalarm is in miliseconds.

File: Database/Buffer/Backend/index.ts [hunk 3]
```diff
@@ -48,10 +52,9 @@ export class Backend {
 	}
 
 	async alarm(): Promise<void> {
-		const archiveTime = 20 * 1000 //TODO: after settlement
 		const now = Date.now()
 		const archiveThreshold = isoly.DateTime.truncate(
-			isoly.DateTime.create(now - archiveTime, "milliseconds"),
+			isoly.DateTime.create(now - this.archiveTime, "milliseconds"),
```
Comment (marcushanses) [inline Database/Buffer/Backend/index.ts:57]: this.archiveTime is in seconds and create() takes miliseconds and now is also in miliseconds.

File: Database/Buffer/Backend/index.ts [hunk 4]
```diff
@@ -61,7 +64,6 @@ export class Backend {
 			return await archivist.reconcile(archiveThreshold)
 		})
 		// TODO: Should be set by some config
```
Comment (marcushanses) [inline Database/Buffer/Backend/index.ts:66]: This comment is literally what you fixed, please remove.

File: Database/Buffer/Backend/index.ts [hunk 5]
```diff
@@ -10,29 +10,33 @@ import { Context } from "./Context"
...
+	private archiveTime: number
+
 	private constructor(private readonly state: DurableObjectState, private environment: Environment) {
 		this.state.blockConcurrencyWhile(async () => {
-			this.partitions = await this.state.storage.get("partitions")
+			this.partitions = (await this.state.storage.get("partitions")) ?? this.partitions
 			!!(await this.state.storage.getAlarm()) ||
 				(await (async () => {
-					await this.state.storage.setAlarm(this.alarmTime)
+					await this.state.storage.setAlarm(this.snooze)
```
Comment (marcushanses) [inline Database/Buffer/Backend/index.ts:23]: This will not work. One can't used undefined to set the alarm. Should also be in miliseconds not in seconds.

**Review by marcushanses — CHANGES_REQUESTED** [review]

**Review by marcushanses — APPROVED** [review]

File: Database/Document.ts [hunk 6]
```diff
@@ -23,4 +23,31 @@ export namespace Document {
 		const { id, created, changed, purged: purged, ...remainder } = document
 		return [{ id, created, changed, purged }, remainder as T]
 	}
+	export function update<T extends Record<string, any> = Document>(original: T, appendee: Partial<T>): T | undefined {
+		return JSON.parse(JSON.stringify({ ...original, ...appendee }))
```
Comment (Magnus93) [inline Database/Document.ts:27]: Why stringify and parse the object?

File: Database/Document.ts [hunk 7]
```diff
@@ -23,4 +23,31 @@ export namespace Document {
...
+	}
+	export function append<T extends Record<string, any> = Document>(
```
Comment (Magnus93) [inline Database/Document.ts:29]: The `append` function could use another test that show how the function adds objects, arrays, etc.

File: Database/Buffer/Backend/update.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+import { router } from "./router"
+
+export async function update(request: http.Request, context: Context): Promise<http.Response.Like | any> {
+	let result: gracely.Result
+	const body = await request.body
+	const amendment = body.amendment
+	const archived = body.archived
+	const storage = context.storage
+	if (!amendment || !amendment.id)
+		result = gracely.client.invalidContent("incomingDocumentument", "An update must be provided.")
```
Comment (Magnus93) [inline Database/Buffer/Backend/update.ts:14]: Weird error

File: Database/Buffer/Backend/append.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+import { router } from "./router"
+
+export async function update(request: http.Request, context: Context): Promise<http.Response.Like | any> {
+	let result: gracely.Result
+	const body = await request.body
+	const amendment = body?.amendment
+	const archived = body?.archived
+	const storage = context.storage
+	if (!amendment || !amendment.id)
+		result = gracely.client.invalidContent("incomingDocumentument", "An update must be provided.")
```
Comment (Magnus93) [inline Database/Buffer/Backend/append.ts:14]: weird error again

**Review by Magnus93 — CHANGES_REQUESTED** [review]

File: Database/Document.ts (hunk 6 above)
Comment (marcushanses) [reply Database/Document.ts:27]: There was a bug and this was my way of making a deep copy :)

**Review by fredrikbryntesson — DISMISSED** [review]

**Review by marcushanses — APPROVED** [review]

File: Database/Document.ts (hunk 7 above)
Comment (marcushanses) [reply Database/Document.ts]: https://github.com/utily/cloudly-storage/pull/42

Comment (Magnus93) [reply Database/Document.ts]: Fantastic

File: Database/Document.ts (hunk 6 above)
Comment (Magnus93) [reply Database/Document.ts]: Nice :D

**Review by Magnus93 — APPROVED** [review]

**Pull request merged**